File: server/tts_worker.py
#!/usr/bin/env python3
"""
Persistent KittenTTS worker — loads the model once and serves synthesis requests over
stdin/stdout as newline-delimited JSON, so app/api/tts/route.ts (via lib/server/ttsWorker.ts)
doesn't pay a full model-reload cost on every message the way the old spawn-per-request
kittentts_adapter.py CLI path did. Mirrors server/stt_worker.py's protocol/shape exactly.

Protocol (line-delimited JSON, flushed after every line so the parent can read as it goes):
  stdin  -> {"id": str, "text": str, "voice": str, "speed": float, "output": <path>}
  stdout -> {"id": str, "output": <path>, "engine": "kittentts"}
         or {"id": str, "error": str}
"""

import sys
import json

from kittentts_adapter import synthesize_file, DEFAULT_VOICE


def main():
    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue

        req_id = None
        try:
            req = json.loads(line)
            req_id = req.get("id")
            result = synthesize_file(
                req["text"],
                req.get("voice") or DEFAULT_VOICE,
                req.get("speed") or 0.9,
                req["output"],
            )
            result["id"] = req_id
        except Exception as e:
            result = {"id": req_id, "error": str(e)}

        print(json.dumps(result), flush=True)


if __name__ == "__main__":
    main()

File: lib/server/ttsWorker.ts
// Manages one persistent Python TTS worker process (server/tts_worker.py) instead of
// spawning a fresh `python kittentts_adapter.py` process per message, which was reloading
// the KittenTTS model from disk on every single request. Mirrors lib/server/sttWorker.ts's
// singleton/protocol shape exactly — see that file for the dev-hot-reload rationale.

import { spawn, ChildProcessWithoutNullStreams } from "node:child_process";
import path from "node:path";
import readline from "node:readline";

interface PendingRequest {
  resolve: (value: { output: string; engine: string }) => void;
  reject: (err: Error) => void;
  timer: NodeJS.Timeout;
}

interface WorkerState {
  proc: ChildProcessWithoutNullStreams;
  pending: Map<string, PendingRequest>;
}

const g = globalThis as unknown as { __ttsWorker?: WorkerState };

function spawnWorker(): WorkerState {
  const scriptPath = path.join(process.cwd(), "server", "tts_worker.py");
  const proc = spawn("python", [scriptPath], { cwd: process.cwd() });
  const pending = new Map<string, PendingRequest>();

  readline.createInterface({ input: proc.stdout }).on("line", (line) => {
    let msg: { id?: string; output?: string; engine?: string; error?: string };
    try {
      msg = JSON.parse(line);
    } catch {
      return;
    }
    if (!msg.id) return;
    const req = pending.get(msg.id);
    if (!req) return;
    pending.delete(msg.id);
    clearTimeout(req.timer);
    if (msg.error) {
      req.reject(new Error(msg.error));
    } else {
      req.resolve({ output: msg.output || "", engine: msg.engine || "kittentts" });
    }
  });

  const onDown = () => {
    if (g.__ttsWorker?.proc === proc) g.__ttsWorker = undefined;
    for (const req of pending.values()) {
      clearTimeout(req.timer);
      req.reject(new Error("TTS worker exited"));
    }
    pending.clear();
  };
  proc.on("exit", onDown);
  proc.on("error", onDown);

  return { proc, pending };
}

function getWorker(): WorkerState {
  if (g.__ttsWorker && !g.__ttsWorker.proc.killed) return g.__ttsWorker;
  g.__ttsWorker = spawnWorker();
  return g.__ttsWorker;
}

export function synthesizeViaWorker(
  text: string,
  voice: string,
  speed: number,
  outputPath: string,
  timeoutMs = 45000
): Promise<{ output: string; engine: string }> {
  const worker = getWorker();
  const id = `${Date.now()}_${Math.random().toString(36).slice(2)}`;

  return new Promise((resolve, reject) => {
    const timer = setTimeout(() => {
      worker.pending.delete(id);
      reject(new Error("TTS worker timed out"));
    }, timeoutMs);

    worker.pending.set(id, { resolve, reject, timer });
    worker.proc.stdin.write(JSON.stringify({ id, text, voice, speed, output: outputPath }) + "\n");
  });
}
